### PR TITLE
add common/promote-kubectl

### PR DESCRIPTION
**Pull request by kalantar**
Signed-off-by: Michael Kalantar <kalantar@us.ibm.com>

Addresses https://github.com/iter8-tools/iter8/issues/798. Doc updates still needed once this is merged.

### CODE REVIEW

**Comment by kalantar** [comment]
Updated with additional option for `-R` flag for `kubectl apply`

**Review by sriumcp — APPROVED** [review]
lgtm

**Comment by huang195** [comment]
@kalantar is there a test yaml file that makes use of `common/kubectl-promote`?

File: lib/common/promote_kubectl.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+// ToBashTask converts a PromoteKubectl task to a Bash task
+func (t *PromoteKubectlTask) ToBashTask() *BashTask {
+	namespace := t.With.Namespace
+	if namespace == nil {
+		ns := viper.GetViper().GetString("experiment_namespace")
+		namespace = &ns
+	}
+
+	script := "kubectl --namespace " + *namespace
```
Comment (huang195) [inline lib/common/promote_kubectl.go:59]: Where is the verb of `kubectl` being passed here?

**Comment by kalantar** [comment]
> @kalantar is there a test yaml file that makes use of `common/kubectl-promote`?

There is not. Tests are written manually creating the task from the go object that would have resulted from a yaml file. See TestMakePromoteKubectlTask

**Comment by huang195** [comment]
So this allows the handler pod to run `kubectl` command on an arbitrary resource file as input, and it will run as `serviceaccount:iter8-controller`. As any k8s resource could be updated here, this service account could require pretty high level of authorization.

**Review by sriumcp — COMMENTED** [review]
Agree with @huang195's comment... on verb.

Also without the `apply` verb, I am not sure what the test is actually testing.

**Review by sriumcp — CHANGES_REQUESTED** [review]
A `Ginkgo` env  test would be useful for this task. See  https://github.com/iter8-tools/handler/blob/main/experiment/experiment_env_test.go for  examples of env tests in this repo.

**Comment by huang195** [comment]
> Also without the apply verb, I am not sure what the test is actually testing.

It's not just `apply`. For K8s resources with `generateName` field, you would need to use `create`. Maybe the verb should be passed as an input.

**Comment by sriumcp** [comment]
> So this allows the handler pod to run kubectl command on an arbitrary resource file as input, and it will run as serviceaccount:iter8-controller. As any k8s resource could be updated here, this service account could require pretty high level of authorization.

serviceaccount:iter8-handler, but the point still applies re: high level of authorization.

Iter8 currently is not multi-tenant. I think this high-level of authorization is ok for this situation. We definitely need to (eventually) answer the multi-tenancy question and design for it.

**Comment by huang195** [comment]
I'm reading the issue linked with this PR, so the example given is:

```
      - task: common/promote-kubectl
        with:
          namespace: test
          manifest: "{{ .promotionManifest }}"
```

but is that really that much simpler than?

```
      - task: common/script
        with: |
          kubectl apply -f {{ .promotionManifest }} -n test
```

This is going a great length to get rid of scripts, even one-liners.

File: lib/common/common_test.go [hunk 2]
```diff
@@ -89,3 +89,25 @@ func TestBashRun(t *testing.T) {
...
+		},
+	})
+	assert.NotEmpty(t, task)
+	assert.NoError(t, err)
+	assert.Equal(t, "default", *task.(*PromoteKubectlTask).With.Namespace)
+	assert.Equal(t, "promote.yaml", task.(*PromoteKubectlTask).With.Manifest)
+	assert.Equal(t, true, *task.(*PromoteKubectlTask).With.Recursive)
+
+	bTask := *task.(*PromoteKubectlTask).ToBashTask()
+	assert.Equal(t, "kubectl --namespace default --recursive --filename promote.yaml", bTask.With.Script)
```
Comment (huang195) [inline lib/common/common_test.go:112]: Missing `kubectl` verb here too?

**Comment by kalantar** [comment]
A gingko test is not suitable f

> A `Ginkgo` env test would be useful for this task. See https://github.com/iter8-tools/handler/blob/main/experiment/experiment_env_test.go for examples of env tests in this repo.

An env test is not suitable for the implementation. They are suitable when calling the go libraries. In this case, `kubectl` is called via a bash task.

**Comment by sriumcp** [comment]
The promote call itself is a go library call. It happens to use `bash` under the covers as an implementation mechanism.

So yes, `Ginkgo` is suitable.

And it would have caught the lack of verb bug in the current implementation.

**Comment by kalantar** [comment]
As a bash  command, kubectl needs a valid out of cluster kube config.  As far as I can tell, this is not available for the test environment that can be created for env tests. I did try.

**Comment by kalantar** [comment]
> So this allows the handler pod to run `kubectl` command on an arbitrary resource file as input, and it will run as `serviceaccount:iter8-controller`. As any k8s resource could be updated here, this service account could require pretty high level of authorization.

Yes. However, we don't make it a requirement that the permission be granted. We have been documenting for each domain the RBAC rules that must be set. Presumably, they will be set for any promote that is defined.

**Review by sriumcp — APPROVED** [review]
lgtm

**Pull request merged**